sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamJavaUdfCalcRule.java

```diff
@@ -52,4 +117,57 @@ public RelNode convert(RelNode rel) {
         RelOptRule.convert(input, input.getTraitSet().replace(BeamLogicalConvention.INSTANCE)),
         calc.getProgram());
   }
+
+  /**
+   * Returns true only if the literal can be correctly implemented by {@link
+   * org.apache.beam.sdk.extensions.sql.impl.rel.BeamCalcRel} in ZetaSQL.
+   */
+  private static boolean udfSupportsLiteral(RexLiteral rexLiteral) {
+    switch (rexLiteral.getType().getSqlTypeName()) {
+      case BIGINT:
+      case BOOLEAN:
+      case DECIMAL:
+      case DOUBLE:
+      case TIME:
+      case TIMESTAMP:
+      case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+      case VARBINARY:
+      case VARCHAR:
+        return true;
+      case DATE: // BEAM-11990
+      default:
+        return false;
+    }
+  }
+
+  /**
+   * Returns true only if the input ref can be correctly implemented by {@link
+   * org.apache.beam.sdk.extensions.sql.impl.rel.BeamCalcRel} in ZetaSQL.
+   */
+  private static boolean udfSupportsInput(RexInputRef rexInputRef) {
```

Comment:
nit: I'm not sure this function adds value, consider just calling `udfSupportsInputType` directly?

Reply:
Done. (Also changed literals for symmetry.)